chainio/txmgr/geometric/geometric.go

```diff
@@ -0,0 +1,623 @@
+func (t *GeometricTxManager) ensureAnyFireblocksTransactionBroadcasted(ctx context.Context, txs []*transaction) error {
+	queryTicker := time.NewTicker(t.params.GetTxReceiptTickerDuration)
+	defer queryTicker.Stop()
+
+	for {
+		for _, tx := range txs {
+			_, err := t.wallet.GetTransactionReceipt(ctx, tx.TxID)
+			if err == nil || errors.Is(err, wallet.ErrReceiptNotYetAvailable) {
+				t.metrics.ObserveBroadcastLatencyMs(time.Since(tx.requestedAt).Milliseconds())
+				return nil
+			}
+		}
+
+		// Wait for the next round.
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-queryTicker.C:
+		}
+	}
+}
+
+// ensureAnyTransactionConfirmed waits until at least one of the transactions is confirmed (mined + confirmationBlocks
+// blocks). It returns the receipt of the first transaction that is confirmed (only one tx can ever be mined given they
+// all have the same nonce).
+func (t *GeometricTxManager) ensureAnyTransactionConfirmed(
+	ctx context.Context,
+	txs []*transaction,
+) (*types.Receipt, error) {
+	queryTicker := time.NewTicker(t.params.GetTxReceiptTickerDuration)
+	defer queryTicker.Stop()
+	var receipt *types.Receipt
+	var err error
+	// transactions that need to be queried. Some transactions will be removed from this map depending on their status.
+	txnsToQuery := make(map[wallet.TxID]*types.Transaction, len(txs))
+	for _, tx := range txs {
+		txnsToQuery[tx.TxID] = tx.Transaction
+	}
+
+	for {
+		for txID, tx := range txnsToQuery {
+			receipt, err = t.wallet.GetTransactionReceipt(ctx, txID)
+			if err == nil {
+				chainTip, err := t.ethClient.BlockNumber(ctx)
+				if err == nil {
+					if receipt.BlockNumber.Uint64()+uint64(t.params.ConfirmationBlocks) > chainTip {
+						t.logger.Debug(
+							"transaction has been mined but don't have enough confirmations at current chain tip",
+							"txnBlockNumber",
+							receipt.BlockNumber.Uint64(),
+							"numConfirmations",
+							t.params.ConfirmationBlocks,
+							"chainTip",
+							chainTip,
+						)
+						break
+					} else {
+						return receipt, nil
+					}
+				} else {
+					t.logger.Debug("failed to get chain tip while waiting for transaction to mine", "err", err)
+				}
+			}
+
+			// TODO(samlaf): how to maintain these better? How do we know which errors to use and where they are
+			// returned from?
+			if errors.Is(err, ethereum.NotFound) || errors.Is(err, wallet.ErrReceiptNotYetAvailable) {
+				// t.logger.Debug("Transaction not yet mined", "nonce", tx.Nonce(), "txID", txID, "txHash",
+				// tx.Hash().Hex(), "err", err)
+			} else if errors.Is(err, wallet.ErrTransactionFailed) {
+				t.logger.Debug("Transaction failed", "txID", txID, "txHash", tx.Hash().Hex(), "err", err)
+				// Remove the transaction from the list of transactions to query.
+				// go seemingly allows deleting from a map while iterating over it:
+				// https://groups.google.com/g/golang-nuts/c/rEmaoxi11_A
+				// although the official spec and faq don't seem to mention this anywhere...
```

Comment:
https://go.dev/doc/effective_go#for

Reply:
This doesn't seem to mention that deleting items while iterating is safe?